### ISSUE TITLE
Eval bug: Gemma-4: SWA checkpoint restoration discards mid-conversation context, causing context regression across turns

### ISSUE THREAD

**Issue by MaineTim** [ggml-org/llama.cpp#21769]
### Name and Version

./build-sycl/bin/llama-server --version
version: 1 (bfd1f45)
built with IntelLLVM 2025.3.3 for Linux x86_64

### Operating systems

Linux

### GGML backends

SYCL

### Hardware

Ryzen 5600XT + Intel ARC Pro B50 (inference only)

### Models

gemma-4-26B-A4B-it-UD-Q6_K_XL

### Problem description & steps to reproduce

On the second turn with the given model, it was clear from its response that it hadn't retained some context. On the second clean run (after reboot), the model generated a similar lack of context response. On the third clean run the model descended into a loop on the second turn. Feeding the logs below to Claude, it generates the following synopsis of the potential cause:
The root cause appears to be in how llama.cpp's sliding window attention (SWA) checkpoint restoration logic handles Gemma-4's hybrid local/global attention architecture.
Gemma-4 interleaves local attention layers (which only need a sliding window of recent tokens, n_swa = 1024 in these logs) with global attention layers (which require the full context history). The checkpoint restoration logic appears to be using the local attention sliding window boundary to determine which checkpoint to restore and where to resume prompt processing for the entire context — including the global attention layers — rather than maintaining full context for the global layers independently.
The observable result is that on turn 2 and beyond, the server restores a checkpoint anchored early in the conversation (positions ~1186-2721 in these logs) and only processes a narrow window of tokens forward from there (~1200-1400 tokens actually evaluated against a full prompt of 11,000-18,000 tokens), silently discarding the intervening conversation. The model then generates a response with no awareness of the middle portion of the conversation.
The bug reproduces consistently across fresh boots, and across different GPU offload configurations, indicating it is not related to memory pressure but is structural to the SWA checkpoint handling for this architecture.

Edited to note that the prompts used on each turn were the same across runs.

### Relevant log output
<details>
<summary>Logs</summary>

```console
srv        update:    - prompt 0x284a27c0:   11549 tokens, checkpoints:  2,  1109.944 MiB
srv  get_availabl: prompt cache update took 560.18 ms
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> ?temp-ext -> dist 
slot launch_slot_: id  3 | task 8828 | processing task, is_child = 0
slot update_slots: id  3 | task 8828 | new prompt, n_ctx_slot = 32768, n_keep = 0, task.n_tokens = 3843
slot update_slots: id  3 | task 8828 | n_past = 2726, slot.prompt.tokens.size() = 11549, seq_id = 3, pos_min = 6941, n_swa = 1024
slot update_slots: id  3 | task 8828 | Checking checkpoint with [0, 2721] against 1702...
slot update_slots: id  3 | task 8828 | restored context checkpoint (pos_min = 0, pos_max = 2721, n_tokens = 2722, n_past = 2721, size = 282.466 MiB)
slot update_slots: id  3 | task 8828 | n_tokens = 2721, memory_seq_rm [2721, end)
srv  log_server_r: done request: POST /v1/chat/completions 127.0.0.1 200
slot update_slots: id  3 | task 8828 | prompt processing progress, n_tokens = 3327, batch.n_tokens = 606, progress = 0.865730
slot update_slots: id  3 | task 8828 | n_tokens = 3327, memory_seq_rm [3327, end)
slot update_slots: id  3 | task 8828 | prompt processing progress, n_tokens = 3839, batch.n_tokens = 512, progress = 0.998959
slot update_slots: id  3 | task 8828 | created context checkpoint 3 of 32 (pos_min = 0, pos_max = 3326, n_tokens = 3327, size = 345.247 MiB)
slot update_slots: id  3 | task 8828 | n_tokens = 3839, memory_seq_rm [3839, end)
slot init_sampler: id  3 | task 8828 | init sampler, took 0.37 ms, tokens: text = 3843, total = 3843
slot update_slots: id  3 | task 8828 | prompt processing done, n_tokens = 3843, batch.n_tokens = 4
slot update_slots: id  3 | task 8828 | created context checkpoint 4 of 32 (pos_min = 0, pos_max = 3838, n_tokens = 3839, size = 398.378 MiB)
```
</details>

[gemma-4-log.txt](https://github.com/user-attachments/files/26646482/gemma-4-log.txt)

[gemma-4-log-2.txt](https://github.com/user-attachments/files/26646484/gemma-4-log-2.txt)

[gemma-4-log-3.txt](https://github.com/user-attachments/files/26646487/gemma-4-log-3.txt)

**Comment by En3Tho** [comment]
Interesting. I've noticed multiple times that model (gemma 4 26a4) behaved like it hadn't got any additinal context and was super-hardcore at following initial prompt. Looks very similar to what you're describing here.

**Comment by MaineTim** [comment]
Exactly what was happening here.

**Comment by MaineTim** [comment]
To follow-up, the symptoms do not appear when the model is loaded with the Vulkan backend. The model is able to retain coherence through multiple turns. So it appears to be limited to the SYCL backend in my configuration.

**Comment by En3Tho** [comment]
I'm using Vulkan backend and I've got into a situation where model repeteatly asked me for "example" and acted like it did not see my latest messages. I will  compare your findings when I hut it again.

**Comment by vizi0uz** [comment]
Enabling `--ctx-checkpoints` changes generation on Gemma 4, reproducibly and from an identical prefill. I do not reproduce the context loss you describe, so this is adjacent evidence rather than a second report of the same symptom.

### Environment

Following the fields of the bug-results template, since this adds a second environment to the report: https://github.com/ggml-org/llama.cpp/blob/master/.github/ISSUE_TEMPLATE/011-bug-results.yml

**Name and version**

```
version: 10219 (c629da565)
built with Clang 20.1.8 for Windows x86_64
```

**Operating systems**: Windows 11.

**GGML backends**: CUDA. Yours is SYCL on Intel ARC, so between us the behaviour is not backend-specific.

**Hardware**: RTX 5070 Ti (16 GiB), Intel Core i5-12400.

**Models**: `gemma-4-12b-it-UD-Q4_K_XL.gguf` with `mmproj-BF16.gguf`, from https://huggingface.co/unsloth/gemma-4-12b-it-GGUF

**Flags**: `--ctx-size 262144 -ngl 99 --flash-attn auto --parallel 1 --cache-type-k q8_0 --cache-type-v q8_0`. The two runs below differ only in `--ctx-checkpoints 0` against `32`.

The template asks for a reproduction through `llama-completion` with `-fit off` where possible. That path is closed here: `llama-completion` in this build rejects `--ctx-checkpoints`, `--swa-checkpoints` and `--cache-ram` with `error: invalid argument`, while accepting `--ctx-size` from the same command line. `llama-cli` does take them, so the flag is not server-only; I have not attempted a reproduction there because the observation below rests on the server's per-request `cache_n` and `prompt_n` to establish that the prefill is byte-for-byte identical across the two configurations, and I have no equivalent instrument in `llama-cli`.

### What I see

Turn 1 of an 8-turn conversation over a 35,684-token document, sampling pinned greedy (`temperature 0`, `top_k 1`, fixed `seed`):

```
ckpt 0   turn 1  cache_n=0  prompt_n=35684  predicted_n=109  sha=05709f37ec77
ckpt 32  turn 1  cache_n=0  prompt_n=35684  predicted_n=217  sha=ec2ffa02c648
```

Same `cache_n`, same `prompt_n`, same prompt, same seed, and a different generation. Nothing has been restored at this point; a checkpoint has only been created. The server log shows a trivial one landing before the large prefill completes:

```
created context checkpoint 1 of 32 (pos_min = 0, pos_max = 7, n_tokens = 8, size = 1.329 MiB)
created context checkpoint 2 of 32 (pos_min = 33632, pos_max = 35167, n_tokens = 35168, size = 170.013 MiB)
```

That is why I do not think this reduces to batching-order noise. I ran each configuration twice on freshly started servers before believing it: `ckpt0-A == ckpt0-B`, `ckpt32-A == ckpt32-B`, `ckpt0 != ckpt32`. Each side is byte-identical to itself, so the divergence is attributable to the flag.

Further in, the two runs separate like this:

```
0 : turn 3 cache_n=35713  prompt_n=29     predicted_n=256   content=empty
32: turn 3 cache_n=35713  prompt_n=29     predicted_n=205   content=7      FANTAIL

0 : turn 7 cache_n=0      prompt_n=35864  predicted_n=256
32: turn 7 cache_n=35840  prompt_n=97     predicted_n=256
```

In this workload checkpoints are the better configuration. Forced re-prefills go from 1 to 0, and several turns produce a visible answer where the `ckpt 0` run spends its whole budget inside the reasoning block and returns `content: ""`. I report that because it suggests the effect changes generation instead of degrading it, which would explain why reports of this differ so much in symptom. On a separate multi-turn retrieval harness the same flag took a run from 8/8 to 7/8, with one answer running away to the token cap. Opposite direction, same underlying instability.

### Minimal repro

Python 3 stdlib, no dependencies. Run it against the same server started twice with the two flag values, then diff the outputs.

Three properties of the workload are load-bearing. The context has to exceed the checkpoint `min spacing` of 8192 tokens or none are created. One turn has to generate past the SWA slack of roughly 512 tokens, which turn 6 requests explicitly so the repro does not depend on the model rambling. And the hash has to cover `reasoning_content`: this model reasons by default, a turn that spends its budget reasoning returns `content: ""`, and hashing `content` alone makes two different runs look identical. My first attempt did exactly that and produced a null result.

```python
#!/usr/bin/env python3
"""Minimal repro: --ctx-checkpoints changes generation output on a SWA model (Gemma 4)."""
import argparse, hashlib, json, os, urllib.request

AP = argparse.ArgumentParser()
AP.add_argument("--port", type=int, default=8082)
AP.add_argument("--host", default="127.0.0.1")
AP.add_argument("--model", default="", help="model id; only needed against a router")
AP.add_argument("--api-key", default=os.environ.get("LLAMA_API_KEY", ""))
AP.add_argument("--lines", type=int, default=1000)
AP.add_argument("--seed", type=int, default=20260801)
A = AP.parse_args()
BASE = f"http://{A.host}:{A.port}"

UNITS = ["node","relay","sensor","collector","aggregator","shard","probe","emitter"]
VERBS = ["completed a scheduled integrity scan","reported nominal intake pressure",
         "synchronised its counters with the upstream reference","rotated its access log to cold storage",
         "cleared a transient parity warning","published its hourly utilisation summary",
         "acknowledged the shift handover","refreshed its interface descriptor table"]
WORDS = ["VELLUM","ROOKERY","FANTAIL","KIRTLE","ZEPHYR","MARLIN","CINDER","QUARRY"]

_s = 12345
def rnd(n):
    global _s
    _s = (1103515245 * _s + 12345) % (1 << 31)
    return _s % n

lines = ["[2026-03-%02d %02d:%02d:%02d] ROUTINE %s %03d %s."
         % (rnd(28)+1, rnd(24), rnd(60), rnd(60), UNITS[rnd(len(UNITS))], rnd(998)+1, VERBS[rnd(len(VERBS))])
         for _ in range(A.lines)]
for t in range(8):
    idx = int(len(lines) * (1 + 92*t/7) / 100.0)
    lines.insert(idx, "[2026-03-15 09:%02d:00] MAINTENANCE cabinet %d: recovery passphrase set to %s."
                 % (t, t+1, WORDS[t]))
HAY = "\n".join(lines)

SHORT = "What is the recovery passphrase for cabinet {n}? Answer with the word only."
LONG  = ("Write a detailed operational report about this log, as at least 40 numbered lines. "
         "Cover the units involved, the kinds of routine events, and the maintenance entries.")
TURNS = [("short",1,256),("short",2,256),("short",3,256),("short",4,256),
         ("short",5,256),("long",None,1500),("short",6,256),("short",7,256)]

def ask(messages, max_tokens):
    body = {"messages": messages, "max_tokens": max_tokens, "temperature": 0, "top_k": 1, "seed": A.seed}
    if A.model: body["model"] = A.model
    req = urllib.request.Request(BASE + "/v1/chat/completions", data=json.dumps(body).encode(),
                                 headers={"Content-Type": "application/json"})
    if A.api_key: req.add_header("Authorization", "Bearer " + A.api_key)
    with urllib.request.urlopen(req, timeout=1800) as r:
        return json.load(r)

msgs = []
print(f"# haystack: {len(lines)} lines")
for i, (kind, n, cap) in enumerate(TURNS):
    q = LONG if kind == "long" else SHORT.format(n=n)
    msgs.append({"role": "user", "content": (HAY + "\n\n" + q) if i == 0 else q})
    r = ask(msgs, cap)
    m = r["choices"][0]["message"]
    content   = m.get("content") or ""
    reasoning = m.get("reasoning_content") or ""
    tim = r.get("timings", {})
    msgs.append({"role": "assistant", "content": content})
    digest = hashlib.sha256((content + "\x00" + reasoning).encode()).hexdigest()[:12]
    print("turn %d %-5s cache_n=%-7s prompt_n=%-7s predicted_n=%-5s content=%-5s sha=%s  %s"
          % (i+1, kind, tim.get("cache_n"), tim.get("prompt_n"), tim.get("predicted_n"),
             "empty" if not content else str(len(content)), digest,
             content.strip().replace("\n", " ")[:40]))
```

Two numbers I ran into on the way that may be useful. A checkpoint on this model costs 170.013 MiB, against the roughly 0.3 MiB quoted in #15293 for its test models; it is the SWA slice, and Gemma 4's is large. At the default of 32 that would exceed the default 8192 MiB `--cache-ram` budget, although `min spacing = 8192` holds the practical count near four at 35k of context. And `/v1/models` is exempt from authentication, answering 200 to an actively invalid bearer token, which cost me a false negative while scripting around this.

A changed generation is not by itself a bug. What I think deserves a look is that it changes from an identical prefill, before any restore, which points at the checkpoint machinery perturbing state it should only be observing.
